addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/filesystem-role-service.js

```diff
@@ -0,0 +1,553 @@
+    ]);
+
+    // Primer
+    // We capture the relations between four items: fs role, study, member account and environment.
+    // We keep the relationship information in three database items across two tables.
+    // 1. Study to fs roles, this is a one-to-many relationship (but we need to make it many-to-many in the future
+    //    if we introduce storage gateway). This relationship is captured in the ResourceUsages table.
+    //    Database item => resource: <studyId>, setName: <studyId>, items: [<fs arn>, ...]
+    // 2. Study to member account to environment using the study. This relationship is also captured in the
+    //    ResourceUsages table.
+    //    Database item => resource: <studyId>, setName: <memberAccountId>, items: [<envId>, ...]
+    // 3. fs role to member account ids. This relationship is captured in the RoleAllocations table.
+    //    Database item => composite key: <accountId, fs role arn>, trust: [<memberAccountId>, ...]
+
+    // The logic
+    // - Get the application role entity from the study.appRoleArn
+    // - Using the resource usage service, get all existing usages of fs roles for the study, see item #1 above.
+    // - For each found fs role arn, load its fs role entity,
+    //   - Does the entity include the member account in its trust and include the study with the same envPermission,
+    //     if so, then we don't need to create an fs role
+    //   - If not, can we fit the member account in the trust policy, if so then update the actual trust document
+    //     in the data source account for the fs role, then update the database, see item #3 above.
+    // - If no fs role entity is found, or the existing on has its trust property full, we need to create a new
+    //   fs role as follows:
+    //   - Create the fs role entity and add the member account to the trust property, see item #3 above.
+    //     - At this point, we don't store the entity in the data base until after we are able to create
+    //       it in the data source account
+    //   - Assume the app role and create the fs role in the data source account
+    //   - Store the entity in the database
+    //   - Add usage of the study by this role, see item #1 above.
+
+    // Get app role entity
+    const appRoleEntity = await appRoleService.mustFind(requestContext, { accountId, bucket, arn: appRoleArn });
+
+    // Get all existing usages by fs roles for this study (keeping env permission in mind)
+    const fsRoleUsages = await resourceUsageService.getResourceUsage(requestContext, {
+      resource: studyResourceId,
+      setName: fsUsageSetName,
+    });
+    const fsRoleEntities = _.get(fsRoleUsages, fsUsageSetName, []);
+
+    // Start a loop to see if existing fs roles for the study have the member account already
+    // or can take it without exceeding the trust doc limit?
+    let fsRoleEntity;
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const fsRoleArn of fsRoleEntities) {
```

Comment:
Out of curiosity, why is the for loop written this way? We can do a `Promise.all` on the first await call to find entities for all fsRole, then use `filter` to find if there's an existing fs role.

Reply:
The logic itself does not lend itself as a filter logic as it contains mutation, multiple decisions and side effects, such as calling mutating AWS IAM APIs.